sensor/common/clusterentities/store_ips.go

```diff
@@ -0,0 +1,244 @@
+	ipMap map[net.IPAddress]set.StringSet
+	// reverseIpMap maps deployment ids to sets of IP addresses associated with this deployment.
+	reverseIPMap map[string]set.FrozenSet[net.IPAddress]
+	// historicalIPs is mimicking ipMap: IP Address -> deploymentID -> historyStatus
+	historicalIPs map[net.IPAddress]map[string]*entityStatus
+}
+
+func newPodIPsStoreWithMemory(numTicks uint16) *podIPsStore {
+	store := &podIPsStore{memorySize: numTicks}
+	concurrency.WithLock(&store.mutex, func() {
+		store.initMapsNoLock()
+	})
+	return store
+}
+
+func (e *podIPsStore) initMapsNoLock() {
+	e.ipMap = make(map[net.IPAddress]set.StringSet)
+	e.reverseIPMap = make(map[string]set.FrozenSet[net.IPAddress])
+	e.historicalIPs = make(map[net.IPAddress]map[string]*entityStatus)
+}
+
+func (e *podIPsStore) resetMaps() {
+	e.mutex.Lock()
+	defer e.mutex.Unlock()
+	// Maps holding historical data must not be wiped on reset! Instead, all entities must be marked as historical.
+	// Must be called before the respective source maps are wiped!
+	// Performance optimization: no need to handle history if history is disabled
+	if !e.historyEnabled() {
+		e.initMapsNoLock()
+		return
+	}
+	for deplID := range e.reverseIPMap {
+		e.moveDeploymentToHistory(deplID)
+	}
+
+	e.ipMap = make(map[net.IPAddress]set.StringSet)
+	e.reverseIPMap = make(map[string]set.FrozenSet[net.IPAddress])
+	e.updateMetricsNoLock()
+}
+
+func (e *podIPsStore) historyEnabled() bool {
+	return e.memorySize > 0
+}
+
+func (e *podIPsStore) updateMetricsNoLock() {
+	metrics.UpdateNumberOfIPs(len(e.ipMap), len(e.historicalIPs))
+}
+
+func (e *podIPsStore) RecordTick() bool {
+	e.mutex.Lock()
+	defer e.mutex.Unlock()
+	removed := false
+	for ip, m := range e.historicalIPs {
+		for deploymentID, status := range m {
+			status.recordTick()
+			// Remove all historical entries that expired in this tick.
+			removed = e.removeFromHistoryIfExpired(deploymentID, ip) || removed
+		}
+	}
+	return removed
+}
+
+func (e *podIPsStore) Apply(updates map[string]*EntityData, incremental bool) {
+	e.mutex.Lock()
+	defer e.mutex.Unlock()
+	e.applyNoLock(updates, incremental)
+}
+
+func (e *podIPsStore) applyNoLock(updates map[string]*EntityData, incremental bool) {
+	defer e.updateMetricsNoLock()
+	if !incremental {
+		for deploymentID := range updates {
+			e.purgeDeploymentNoLock(deploymentID)
+		}
+	}
+	for deploymentID, data := range updates {
+		if data.isDeleteOnly() {
+			// A call to Apply() with empty payload of the updates map (no values) is meant to be a delete operation.
+			continue
+		}
+		e.applySingleNoLock(deploymentID, *data)
+	}
+}
+
+func (e *podIPsStore) purgeDeploymentNoLock(deploymentID string) {
+	if e.historyEnabled() {
+		e.moveDeploymentToHistory(deploymentID)
+	} else {
+		e.deleteDeploymentFromCurrent(deploymentID)
+		// In case we allow in the future to disable history during runtime, we would need to remove here all
+		// expired data for deploymentID from the history.
```

Comment:
Wouldn't it make more sense to delete everything from the history map when disabling history during runtime? And then nothing would need to be done here.

Reply:
> I am thinking maybe we should drop the assumption that history may be enabled/disabled during runtime? WDYT @lvalerom ?

I'm not sure if there's a good reason to disable the history except as a fail-safe in case of a bug but I'm a little bit out of the loop when it comes to the runtime configuration.

Reply:
There will be a follow-up PR for this part - resolving the comment for now.